<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.4.553">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Anna Nauruschat">

<title>R scripts for: “Hybrid genome assemblies of R. sphaeroides DSM158 and its substrain H2”</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="Supplementary_R_plots_files/libs/clipboard/clipboard.min.js"></script>
<script src="Supplementary_R_plots_files/libs/quarto-html/quarto.js"></script>
<script src="Supplementary_R_plots_files/libs/quarto-html/popper.min.js"></script>
<script src="Supplementary_R_plots_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="Supplementary_R_plots_files/libs/quarto-html/anchor.min.js"></script>
<link href="Supplementary_R_plots_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="Supplementary_R_plots_files/libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="Supplementary_R_plots_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="Supplementary_R_plots_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="Supplementary_R_plots_files/libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">


</head>

<body class="fullcontent">

<div id="quarto-content" class="page-columns page-rows-contents page-layout-article">

<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">R scripts for: “Hybrid genome assemblies of R. sphaeroides DSM158 and its substrain H2”</h1>
</div>



<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Anna Nauruschat </p>
          </div>
  </div>
    
  
    
  </div>
  


</header>


<section id="circos-plot-of-the-assemblies-using-circlize" class="level1">
<h1>Circos plot of the assemblies using circlize</h1>
<section id="loading-the-required-packages" class="level4">
<h4 class="anchored" data-anchor-id="loading-the-required-packages">Loading the required packages</h4>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(Biostrings)     <span class="co"># version 2.70.3</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: BiocGenerics</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>
Attaching package: 'BiocGenerics'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following objects are masked from 'package:stats':

    IQR, mad, sd, var, xtabs</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following objects are masked from 'package:base':

    anyDuplicated, aperm, append, as.data.frame, basename, cbind,
    colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,
    get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,
    match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,
    Position, rank, rbind, Reduce, rownames, sapply, setdiff, sort,
    table, tapply, union, unique, unsplit, which.max, which.min</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: S4Vectors</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: stats4</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>
Attaching package: 'S4Vectors'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following object is masked from 'package:utils':

    findMatches</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following objects are masked from 'package:base':

    expand.grid, I, unname</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: IRanges</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: XVector</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: GenomeInfoDb</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>
Attaching package: 'Biostrings'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following object is masked from 'package:base':

    strsplit</code></pre>
</div>
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(rtracklayer)    <span class="co"># version 1.62.0</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: GenomicRanges</code></pre>
</div>
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(GenomicRanges)  <span class="co"># version 1.54.1</span></span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(circlize)       <span class="co"># version 0.4.16</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>========================================
circlize version 0.4.16
CRAN page: https://cran.r-project.org/package=circlize
Github page: https://github.com/jokergoo/circlize
Documentation: https://jokergoo.github.io/circlize_book/book/

If you use it in published research, please cite:
Gu, Z. circlize implements and enhances circular visualization
  in R. Bioinformatics 2014.

This message can be suppressed by:
  suppressPackageStartupMessages(library(circlize))
========================================</code></pre>
</div>
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(data.table)     <span class="co"># version 1.17.8</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>
Attaching package: 'data.table'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following object is masked from 'package:GenomicRanges':

    shift</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following object is masked from 'package:IRanges':

    shift</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following objects are masked from 'package:S4Vectors':

    first, second</code></pre>
</div>
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(RColorBrewer)   <span class="co"># version 1.1-3</span></span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggsci)          <span class="co"># version 4.0.0</span></span>
<span id="cb25-3"><a href="#cb25-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(scales)         </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="import-the-genomes-and-the-genome-annotations" class="level4">
<h4 class="anchored" data-anchor-id="import-the-genomes-and-the-genome-annotations">Import the genomes and the genome annotations</h4>
<div class="cell">
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a>genomeA <span class="ot">&lt;-</span> <span class="fu">readDNAStringSet</span>(<span class="st">"/home/Drives/HDD03_06T_SDE/anna/SyntenyPlotDSMSubH2/DSM158/Cereibacter_sphaeroides_reoriented.fasta"</span>)</span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a>gffA <span class="ot">&lt;-</span> <span class="fu">import</span>(<span class="st">"/home/Drives/HDD03_06T_SDE/anna/SyntenyPlotDSMSubH2/DSM158/annot.gff"</span>)</span>
<span id="cb26-3"><a href="#cb26-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-4"><a href="#cb26-4" aria-hidden="true" tabindex="-1"></a>genomeB <span class="ot">&lt;-</span> <span class="fu">readDNAStringSet</span>(<span class="st">"/home/Drives/HDD03_06T_SDE/anna/SyntenyPlotDSMSubH2/SubH2/GCF_049434525.1_MWCSPHH2ANNA_genomic.fasta"</span>)</span>
<span id="cb26-5"><a href="#cb26-5" aria-hidden="true" tabindex="-1"></a>gffB <span class="ot">&lt;-</span> <span class="fu">import</span>(<span class="st">"/home/Drives/HDD03_06T_SDE/anna/SyntenyPlotDSMSubH2/SubH2/genomic.gff"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="tidying-the-sequence-names-remove-file-extension-and-rename-the-chromosomes" class="level4">
<h4 class="anchored" data-anchor-id="tidying-the-sequence-names-remove-file-extension-and-rename-the-chromosomes">Tidying the sequence names (remove file extension) and rename the chromosomes</h4>
<div class="cell">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a><span class="co">#remove the file extensions (.fasta or .gff) from the sequence names</span></span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(genomeA) <span class="ot">&lt;-</span> <span class="fu">sub</span>(<span class="st">" .*"</span>, <span class="st">""</span>, <span class="fu">names</span>(genomeA))</span>
<span id="cb27-3"><a href="#cb27-3" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(genomeB) <span class="ot">&lt;-</span> <span class="fu">sub</span>(<span class="st">" .*"</span>, <span class="st">""</span>, <span class="fu">names</span>(genomeB))</span>
<span id="cb27-4"><a href="#cb27-4" aria-hidden="true" tabindex="-1"></a><span class="fu">seqlevels</span>(gffA) <span class="ot">&lt;-</span> <span class="fu">sub</span>(<span class="st">" .*"</span>, <span class="st">""</span>, <span class="fu">seqlevels</span>(gffA))</span>
<span id="cb27-5"><a href="#cb27-5" aria-hidden="true" tabindex="-1"></a><span class="fu">seqlevels</span>(gffB) <span class="ot">&lt;-</span> <span class="fu">sub</span>(<span class="st">" .*"</span>, <span class="st">""</span>, <span class="fu">seqlevels</span>(gffB))</span>
<span id="cb27-6"><a href="#cb27-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-7"><a href="#cb27-7" aria-hidden="true" tabindex="-1"></a><span class="co">#Rename the chromosomes A_... or B_...</span></span>
<span id="cb27-8"><a href="#cb27-8" aria-hidden="true" tabindex="-1"></a>new_namesA <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">"A_"</span>, <span class="fu">names</span>(genomeA))</span>
<span id="cb27-9"><a href="#cb27-9" aria-hidden="true" tabindex="-1"></a>new_namesB <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">"B_"</span>, <span class="fu">names</span>(genomeB))</span>
<span id="cb27-10"><a href="#cb27-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-11"><a href="#cb27-11" aria-hidden="true" tabindex="-1"></a><span class="co">#Extract the lengths for each chromosome</span></span>
<span id="cb27-12"><a href="#cb27-12" aria-hidden="true" tabindex="-1"></a>chrom_lensA <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">chr=</span>new_namesA, <span class="at">start=</span><span class="dv">0</span>, <span class="at">end=</span><span class="fu">width</span>(genomeA))</span>
<span id="cb27-13"><a href="#cb27-13" aria-hidden="true" tabindex="-1"></a>chrom_lensB <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">chr=</span>new_namesB, <span class="at">start=</span><span class="dv">0</span>, <span class="at">end=</span><span class="fu">width</span>(genomeB))</span>
<span id="cb27-14"><a href="#cb27-14" aria-hidden="true" tabindex="-1"></a>chrom_lens <span class="ot">&lt;-</span> <span class="fu">rbind</span>(chrom_lensA, chrom_lensB)</span>
<span id="cb27-15"><a href="#cb27-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-16"><a href="#cb27-16" aria-hidden="true" tabindex="-1"></a><span class="co">#Initialize the sectors for circlize</span></span>
<span id="cb27-17"><a href="#cb27-17" aria-hidden="true" tabindex="-1"></a>chrom_lens<span class="sc">$</span>chr <span class="ot">&lt;-</span> <span class="fu">as.character</span>(chrom_lens<span class="sc">$</span>chr)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="create-a-sliding-window-for-the-calculation-of-gene-densities-and-gc-content-of-size-10000-bp" class="level4">
<h4 class="anchored" data-anchor-id="create-a-sliding-window-for-the-calculation-of-gene-densities-and-gc-content-of-size-10000-bp">Create a sliding window (for the calculation of gene densities and GC content) of size 10,000 bp</h4>
<div class="cell">
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a>win_size <span class="ot">&lt;-</span> <span class="dv">10000</span></span>
<span id="cb28-2"><a href="#cb28-2" aria-hidden="true" tabindex="-1"></a>windowsA <span class="ot">&lt;-</span> <span class="fu">tileGenome</span>(<span class="fu">seqlengths</span>(genomeA), <span class="at">tilewidth=</span>win_size, <span class="at">cut.last.tile.in.chrom=</span><span class="cn">TRUE</span>)</span>
<span id="cb28-3"><a href="#cb28-3" aria-hidden="true" tabindex="-1"></a>windowsB <span class="ot">&lt;-</span> <span class="fu">tileGenome</span>(<span class="fu">seqlengths</span>(genomeB), <span class="at">tilewidth=</span>win_size, <span class="at">cut.last.tile.in.chrom=</span><span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="calculate-the-gene-densities" class="level4">
<h4 class="anchored" data-anchor-id="calculate-the-gene-densities">Calculate the gene densities</h4>
<div class="cell">
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a>genesA <span class="ot">&lt;-</span> gffA[gffA<span class="sc">$</span>type<span class="sc">==</span><span class="st">"gene"</span>]</span>
<span id="cb29-2"><a href="#cb29-2" aria-hidden="true" tabindex="-1"></a>gene_countsA <span class="ot">&lt;-</span> <span class="fu">countOverlaps</span>(windowsA, genesA)</span>
<span id="cb29-3"><a href="#cb29-3" aria-hidden="true" tabindex="-1"></a>gendichteA <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(</span>
<span id="cb29-4"><a href="#cb29-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">chr=</span><span class="fu">factor</span>(<span class="fu">paste0</span>(<span class="st">"A_"</span>, <span class="fu">seqnames</span>(windowsA)), <span class="at">levels=</span>new_namesA),</span>
<span id="cb29-5"><a href="#cb29-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">start=</span><span class="fu">start</span>(windowsA), <span class="at">end=</span><span class="fu">end</span>(windowsA), <span class="at">value=</span>gene_countsA</span>
<span id="cb29-6"><a href="#cb29-6" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb29-7"><a href="#cb29-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb29-8"><a href="#cb29-8" aria-hidden="true" tabindex="-1"></a>genesB <span class="ot">&lt;-</span> gffB[gffB<span class="sc">$</span>type<span class="sc">==</span><span class="st">"gene"</span>]</span>
<span id="cb29-9"><a href="#cb29-9" aria-hidden="true" tabindex="-1"></a>gene_countsB <span class="ot">&lt;-</span> <span class="fu">countOverlaps</span>(windowsB, genesB)</span>
<span id="cb29-10"><a href="#cb29-10" aria-hidden="true" tabindex="-1"></a>gendichteB <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(</span>
<span id="cb29-11"><a href="#cb29-11" aria-hidden="true" tabindex="-1"></a>  <span class="at">chr=</span><span class="fu">factor</span>(<span class="fu">paste0</span>(<span class="st">"B_"</span>, <span class="fu">seqnames</span>(windowsB)), <span class="at">levels=</span>new_namesB),</span>
<span id="cb29-12"><a href="#cb29-12" aria-hidden="true" tabindex="-1"></a>  <span class="at">start=</span><span class="fu">start</span>(windowsB), <span class="at">end=</span><span class="fu">end</span>(windowsB), <span class="at">value=</span>gene_countsB</span>
<span id="cb29-13"><a href="#cb29-13" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="calculate-the-gc-content" class="level4">
<h4 class="anchored" data-anchor-id="calculate-the-gc-content">Calculate the GC content</h4>
<div class="cell">
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a>calc_gc <span class="ot">&lt;-</span> <span class="cf">function</span>(genome, windows, prefix) {</span>
<span id="cb30-2"><a href="#cb30-2" aria-hidden="true" tabindex="-1"></a>  vals <span class="ot">&lt;-</span> <span class="fu">numeric</span>(<span class="fu">length</span>(windows))</span>
<span id="cb30-3"><a href="#cb30-3" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span> (i <span class="cf">in</span> <span class="fu">seq_along</span>(windows)) {</span>
<span id="cb30-4"><a href="#cb30-4" aria-hidden="true" tabindex="-1"></a>    chr <span class="ot">&lt;-</span> <span class="fu">as.character</span>(<span class="fu">seqnames</span>(windows[i]))</span>
<span id="cb30-5"><a href="#cb30-5" aria-hidden="true" tabindex="-1"></a>    seq <span class="ot">&lt;-</span> <span class="fu">subseq</span>(genome[[chr]], <span class="fu">start</span>(windows[i]), <span class="fu">end</span>(windows[i]))</span>
<span id="cb30-6"><a href="#cb30-6" aria-hidden="true" tabindex="-1"></a>    vals[i] <span class="ot">&lt;-</span> <span class="fu">sum</span>(<span class="fu">letterFrequency</span>(seq, <span class="fu">c</span>(<span class="st">"G"</span>,<span class="st">"C"</span>), <span class="at">as.prob=</span><span class="cn">TRUE</span>))</span>
<span id="cb30-7"><a href="#cb30-7" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb30-8"><a href="#cb30-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">data.frame</span>(</span>
<span id="cb30-9"><a href="#cb30-9" aria-hidden="true" tabindex="-1"></a>    <span class="at">chr=</span><span class="fu">factor</span>(<span class="fu">paste0</span>(prefix, <span class="fu">seqnames</span>(windows)), <span class="at">levels=</span><span class="fu">c</span>(new_namesA,new_namesB)),</span>
<span id="cb30-10"><a href="#cb30-10" aria-hidden="true" tabindex="-1"></a>    <span class="at">start=</span><span class="fu">start</span>(windows), <span class="at">end=</span><span class="fu">end</span>(windows), <span class="at">value=</span>vals</span>
<span id="cb30-11"><a href="#cb30-11" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb30-12"><a href="#cb30-12" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb30-13"><a href="#cb30-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-14"><a href="#cb30-14" aria-hidden="true" tabindex="-1"></a>gcA <span class="ot">&lt;-</span> <span class="fu">calc_gc</span>(genomeA, windowsA, <span class="st">"A_"</span>)</span>
<span id="cb30-15"><a href="#cb30-15" aria-hidden="true" tabindex="-1"></a>gcB <span class="ot">&lt;-</span> <span class="fu">calc_gc</span>(genomeB, windowsB, <span class="st">"B_"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="data-for-the-phage-sequences-phaster-web-server" class="level4">
<h4 class="anchored" data-anchor-id="data-for-the-phage-sequences-phaster-web-server">Data for the phage sequences (PHASTER web-server)</h4>
<div class="cell">
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a>repeatA <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(</span>
<span id="cb31-2"><a href="#cb31-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">chr =</span> <span class="fu">c</span>(<span class="st">"1"</span>,<span class="st">"1"</span>,<span class="st">"1"</span>,<span class="st">"1"</span>, <span class="st">"1"</span>, <span class="st">"2"</span>, <span class="st">"2"</span>, <span class="st">"2"</span>),</span>
<span id="cb31-3"><a href="#cb31-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">start =</span> <span class="fu">c</span>(<span class="dv">299942</span>, <span class="dv">312405</span>, <span class="dv">714792</span>, <span class="dv">1045831</span>, <span class="dv">1184739</span>, <span class="dv">16443</span>, <span class="dv">36955</span>, <span class="dv">184600</span>),</span>
<span id="cb31-4"><a href="#cb31-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">end   =</span> <span class="fu">c</span>(<span class="dv">314808</span>, <span class="dv">334525</span>, <span class="dv">756993</span>, <span class="dv">1067027</span>, <span class="dv">1202250</span>, <span class="dv">38142</span>, <span class="dv">54813</span>, <span class="dv">208162</span>),</span>
<span id="cb31-5"><a href="#cb31-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">type  =</span> <span class="fu">c</span>(<span class="st">"Incomplete"</span>,<span class="st">"Incomplete"</span>,<span class="st">"Questionable"</span>,<span class="st">"Incomplete"</span>, <span class="st">"Incomplete"</span>,<span class="st">"Incomplete"</span>, <span class="st">"Incomplete"</span>,<span class="st">"Incomplete"</span>)</span>
<span id="cb31-6"><a href="#cb31-6" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb31-7"><a href="#cb31-7" aria-hidden="true" tabindex="-1"></a>repeatB <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(</span>
<span id="cb31-8"><a href="#cb31-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">chr =</span> <span class="fu">c</span>(<span class="st">"NZ_CP186561.1"</span>,<span class="st">"NZ_CP186561.1"</span>,<span class="st">"NZ_CP186561.1"</span>,<span class="st">"NZ_CP186561.1"</span>, <span class="st">"NZ_CP186561.1"</span>, <span class="st">"NZ_CP186562.1"</span>, <span class="st">"NZ_CP186562.1"</span>, <span class="st">"NZ_CP186562.1"</span>),</span>
<span id="cb31-9"><a href="#cb31-9" aria-hidden="true" tabindex="-1"></a>  <span class="at">start =</span> <span class="fu">c</span>(<span class="dv">299941</span>, <span class="dv">312404</span>, <span class="dv">714790</span>, <span class="dv">1045832</span>, <span class="dv">1184740</span>, <span class="dv">9996</span>, <span class="dv">157638</span>, <span class="dv">920625</span>),</span>
<span id="cb31-10"><a href="#cb31-10" aria-hidden="true" tabindex="-1"></a>  <span class="at">end   =</span> <span class="fu">c</span>(<span class="dv">314807</span>, <span class="dv">333081</span>, <span class="dv">756989</span>, <span class="dv">1067028</span>, <span class="dv">1202251</span>, <span class="dv">27853</span>, <span class="dv">181200</span>, <span class="dv">944010</span>),</span>
<span id="cb31-11"><a href="#cb31-11" aria-hidden="true" tabindex="-1"></a>  <span class="at">type  =</span> <span class="fu">c</span>(<span class="st">"Incomplete"</span>,<span class="st">"Incomplete"</span>,<span class="st">"Questionable"</span>,<span class="st">"Incomplete"</span>, <span class="st">"Incomplete"</span>,<span class="st">"Incomplete"</span>, <span class="st">"Incomplete"</span>,<span class="st">"Incomplete"</span>)</span>
<span id="cb31-12"><a href="#cb31-12" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb31-13"><a href="#cb31-13" aria-hidden="true" tabindex="-1"></a>repeatA<span class="sc">$</span>chr <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">"A_"</span>, repeatA<span class="sc">$</span>chr)</span>
<span id="cb31-14"><a href="#cb31-14" aria-hidden="true" tabindex="-1"></a>repeatB<span class="sc">$</span>chr <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">"B_"</span>, repeatB<span class="sc">$</span>chr)</span>
<span id="cb31-15"><a href="#cb31-15" aria-hidden="true" tabindex="-1"></a>repeat_df <span class="ot">&lt;-</span> <span class="fu">rbind</span>(repeatA, repeatB)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="read-the-coordinates-from-a-mummer-coordinates-file-for-the-synteny-track" class="level4">
<h4 class="anchored" data-anchor-id="read-the-coordinates-from-a-mummer-coordinates-file-for-the-synteny-track">Read the coordinates from a MUMMer coordinates file (for the synteny track)</h4>
<div class="cell">
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a>read_mummer_coords <span class="ot">&lt;-</span> <span class="cf">function</span>(file){</span>
<span id="cb32-2"><a href="#cb32-2" aria-hidden="true" tabindex="-1"></a>  raw <span class="ot">&lt;-</span> <span class="fu">readLines</span>(file)</span>
<span id="cb32-3"><a href="#cb32-3" aria-hidden="true" tabindex="-1"></a>  header_line <span class="ot">&lt;-</span> <span class="fu">grep</span>(<span class="st">"</span><span class="sc">\\</span><span class="st">[S1</span><span class="sc">\\</span><span class="st">]"</span>, raw)</span>
<span id="cb32-4"><a href="#cb32-4" aria-hidden="true" tabindex="-1"></a>  data_lines <span class="ot">&lt;-</span> raw[(header_line<span class="sc">+</span><span class="dv">2</span>)<span class="sc">:</span><span class="fu">length</span>(raw)]</span>
<span id="cb32-5"><a href="#cb32-5" aria-hidden="true" tabindex="-1"></a>  data_clean <span class="ot">&lt;-</span> <span class="fu">gsub</span>(<span class="st">"</span><span class="sc">\\</span><span class="st">|"</span>, <span class="st">""</span>, data_lines)</span>
<span id="cb32-6"><a href="#cb32-6" aria-hidden="true" tabindex="-1"></a>  data_clean <span class="ot">&lt;-</span> <span class="fu">gsub</span>(<span class="st">"</span><span class="sc">\\</span><span class="st">s+"</span>, <span class="st">" "</span>, data_clean)</span>
<span id="cb32-7"><a href="#cb32-7" aria-hidden="true" tabindex="-1"></a>  coords <span class="ot">&lt;-</span> <span class="fu">read.table</span>(<span class="at">text=</span>data_clean, <span class="at">header=</span><span class="cn">FALSE</span>, <span class="at">stringsAsFactors=</span><span class="cn">FALSE</span>)</span>
<span id="cb32-8"><a href="#cb32-8" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span>(<span class="fu">ncol</span>(coords)<span class="sc">==</span><span class="dv">8</span>){</span>
<span id="cb32-9"><a href="#cb32-9" aria-hidden="true" tabindex="-1"></a>    <span class="fu">colnames</span>(coords) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"startA"</span>,<span class="st">"endA"</span>,<span class="st">"startB"</span>,<span class="st">"endB"</span>,<span class="st">"lenA"</span>,<span class="st">"lenB"</span>,<span class="st">"identity"</span>,<span class="st">"chrA"</span>)</span>
<span id="cb32-10"><a href="#cb32-10" aria-hidden="true" tabindex="-1"></a>    coords<span class="sc">$</span>chrB <span class="ot">&lt;-</span> coords<span class="sc">$</span>chrA</span>
<span id="cb32-11"><a href="#cb32-11" aria-hidden="true" tabindex="-1"></a>  } <span class="cf">else</span> <span class="cf">if</span>(<span class="fu">ncol</span>(coords)<span class="sc">==</span><span class="dv">11</span>){</span>
<span id="cb32-12"><a href="#cb32-12" aria-hidden="true" tabindex="-1"></a>    <span class="fu">colnames</span>(coords) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"startA"</span>,<span class="st">"endA"</span>,<span class="st">"startB"</span>,<span class="st">"endB"</span>,</span>
<span id="cb32-13"><a href="#cb32-13" aria-hidden="true" tabindex="-1"></a>                          <span class="st">"lenA"</span>,<span class="st">"lenB"</span>,<span class="st">"identity"</span>,</span>
<span id="cb32-14"><a href="#cb32-14" aria-hidden="true" tabindex="-1"></a>                          <span class="st">"lenRef"</span>,<span class="st">"lenQry"</span>,<span class="st">"chrA"</span>,<span class="st">"chrB"</span>)</span>
<span id="cb32-15"><a href="#cb32-15" aria-hidden="true" tabindex="-1"></a>  } <span class="cf">else</span> <span class="cf">if</span>(<span class="fu">ncol</span>(coords)<span class="sc">==</span><span class="dv">13</span>){</span>
<span id="cb32-16"><a href="#cb32-16" aria-hidden="true" tabindex="-1"></a>    <span class="fu">colnames</span>(coords) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"startA"</span>,<span class="st">"endA"</span>,<span class="st">"startB"</span>,<span class="st">"endB"</span>,</span>
<span id="cb32-17"><a href="#cb32-17" aria-hidden="true" tabindex="-1"></a>                          <span class="st">"lenA"</span>,<span class="st">"lenB"</span>,<span class="st">"identity"</span>,</span>
<span id="cb32-18"><a href="#cb32-18" aria-hidden="true" tabindex="-1"></a>                          <span class="st">"lenRef"</span>,<span class="st">"lenQry"</span>,<span class="st">"covRef"</span>,<span class="st">"covQry"</span>,<span class="st">"chrA"</span>,<span class="st">"chrB"</span>)</span>
<span id="cb32-19"><a href="#cb32-19" aria-hidden="true" tabindex="-1"></a>  } <span class="cf">else</span> {<span class="fu">stop</span>(<span class="st">"Unbekanntes Format!"</span>)}</span>
<span id="cb32-20"><a href="#cb32-20" aria-hidden="true" tabindex="-1"></a>  coords</span>
<span id="cb32-21"><a href="#cb32-21" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb32-22"><a href="#cb32-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-23"><a href="#cb32-23" aria-hidden="true" tabindex="-1"></a>coords <span class="ot">&lt;-</span> <span class="fu">read_mummer_coords</span>(<span class="st">"/home/Drives/HDD03_06T_SDE/anna/SyntenyPlotDSMSubH2/MUMer/synteny.coords"</span>)</span>
<span id="cb32-24"><a href="#cb32-24" aria-hidden="true" tabindex="-1"></a>synteny <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(</span>
<span id="cb32-25"><a href="#cb32-25" aria-hidden="true" tabindex="-1"></a>  <span class="at">chrA =</span> <span class="fu">paste0</span>(<span class="st">"A_"</span>, coords<span class="sc">$</span>chrA),</span>
<span id="cb32-26"><a href="#cb32-26" aria-hidden="true" tabindex="-1"></a>  <span class="at">startA =</span> coords<span class="sc">$</span>startA,</span>
<span id="cb32-27"><a href="#cb32-27" aria-hidden="true" tabindex="-1"></a>  <span class="at">endA =</span> coords<span class="sc">$</span>endA,</span>
<span id="cb32-28"><a href="#cb32-28" aria-hidden="true" tabindex="-1"></a>  <span class="at">chrB =</span> <span class="fu">paste0</span>(<span class="st">"B_"</span>, coords<span class="sc">$</span>chrB),</span>
<span id="cb32-29"><a href="#cb32-29" aria-hidden="true" tabindex="-1"></a>  <span class="at">startB =</span> coords<span class="sc">$</span>startB,</span>
<span id="cb32-30"><a href="#cb32-30" aria-hidden="true" tabindex="-1"></a>  <span class="at">endB =</span> coords<span class="sc">$</span>endB</span>
<span id="cb32-31"><a href="#cb32-31" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="color-palette-for-the-links-npg-palette-from-the-ggsci-package-and-for-the-other-tracks" class="level4">
<h4 class="anchored" data-anchor-id="color-palette-for-the-links-npg-palette-from-the-ggsci-package-and-for-the-other-tracks">Color palette for the links (npg palette from the <code>ggsci</code> package) and for the other tracks</h4>
<div class="cell">
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a>link_colors <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="fu">pal_npg</span>(<span class="st">"nrc"</span>, <span class="at">alpha =</span> <span class="fl">0.2</span>)(<span class="dv">10</span>), </span>
<span id="cb33-2"><a href="#cb33-2" aria-hidden="true" tabindex="-1"></a>                 <span class="fu">alpha</span>(<span class="st">"#00d087B2"</span>,<span class="fl">0.2</span>), </span>
<span id="cb33-3"><a href="#cb33-3" aria-hidden="true" tabindex="-1"></a>                 <span class="fu">alpha</span>(<span class="st">"#0c5f88b2"</span>,<span class="fl">0.2</span>)</span>
<span id="cb33-4"><a href="#cb33-4" aria-hidden="true" tabindex="-1"></a>                 )</span>
<span id="cb33-5"><a href="#cb33-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb33-6"><a href="#cb33-6" aria-hidden="true" tabindex="-1"></a>colorA <span class="ot">&lt;-</span> <span class="st">"#00a087B2"</span>  <span class="co"># Green for genome A (DSM158)</span></span>
<span id="cb33-7"><a href="#cb33-7" aria-hidden="true" tabindex="-1"></a>colorB <span class="ot">&lt;-</span> <span class="st">"#0c5488b2"</span>  <span class="co"># Blue for genome B (SubH2)</span></span>
<span id="cb33-8"><a href="#cb33-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb33-9"><a href="#cb33-9" aria-hidden="true" tabindex="-1"></a><span class="co"># match genomes wth colors</span></span>
<span id="cb33-10"><a href="#cb33-10" aria-hidden="true" tabindex="-1"></a>chrom_colors <span class="ot">&lt;-</span> <span class="fu">ifelse</span>(<span class="fu">grepl</span>(<span class="st">"^A_"</span>, chrom_lens<span class="sc">$</span>chr), colorA, colorB)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="prepare-the-new-chromosome-names" class="level4">
<h4 class="anchored" data-anchor-id="prepare-the-new-chromosome-names">Prepare the new chromosome names</h4>
<div class="cell">
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Chromosome-IDs in the plot</span></span>
<span id="cb34-2"><a href="#cb34-2" aria-hidden="true" tabindex="-1"></a>chrom_ids <span class="ot">&lt;-</span> chrom_lens<span class="sc">$</span>chr</span>
<span id="cb34-3"><a href="#cb34-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-4"><a href="#cb34-4" aria-hidden="true" tabindex="-1"></a><span class="co"># New names (ordered)</span></span>
<span id="cb34-5"><a href="#cb34-5" aria-hidden="true" tabindex="-1"></a>new_labels <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"Chromosome 1"</span>, <span class="st">"Chromosome 2"</span>, <span class="st">"PlA"</span>,</span>
<span id="cb34-6"><a href="#cb34-6" aria-hidden="true" tabindex="-1"></a>                <span class="st">"PlB"</span>, <span class="st">"PlC"</span>, <span class="st">"PlDx"</span>,<span class="st">"Chromosome 1"</span>,</span>
<span id="cb34-7"><a href="#cb34-7" aria-hidden="true" tabindex="-1"></a>                <span class="st">"Chromosome 2"</span>, <span class="st">"PlA"</span>, <span class="st">"PlB"</span>, <span class="st">"PlC"</span>, <span class="st">"PlDx"</span>)</span>
<span id="cb34-8"><a href="#cb34-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-9"><a href="#cb34-9" aria-hidden="true" tabindex="-1"></a><span class="co"># Control if there are the same number og new labels and old labels</span></span>
<span id="cb34-10"><a href="#cb34-10" aria-hidden="true" tabindex="-1"></a><span class="fu">stopifnot</span>(<span class="fu">length</span>(chrom_ids) <span class="sc">==</span> <span class="fu">length</span>(new_labels))</span>
<span id="cb34-11"><a href="#cb34-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-12"><a href="#cb34-12" aria-hidden="true" tabindex="-1"></a><span class="co"># Match new labels and old labels</span></span>
<span id="cb34-13"><a href="#cb34-13" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(new_labels) <span class="ot">&lt;-</span> chrom_ids</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="circos-plot" class="level4">
<h4 class="anchored" data-anchor-id="circos-plot">Circos plot</h4>
<div class="cell">
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a><span class="fu">circos.clear</span>()</span>
<span id="cb35-2"><a href="#cb35-2" aria-hidden="true" tabindex="-1"></a><span class="fu">circos.par</span>(<span class="st">"start.degree"</span><span class="ot">=</span><span class="dv">90</span>, <span class="at">gap.degree=</span><span class="fu">c</span>(<span class="fu">rep</span>(<span class="dv">3</span>,<span class="dv">5</span>),<span class="dv">6</span>,<span class="fu">rep</span>(<span class="dv">3</span>,<span class="dv">5</span>),<span class="dv">6</span>), <span class="at">cell.padding =</span> <span class="fu">c</span>(<span class="fl">0.02</span>, <span class="dv">0</span>, <span class="fl">0.02</span>, <span class="dv">0</span>), <span class="at">track.margin=</span><span class="fu">c</span>(<span class="fl">0.01</span>,<span class="fl">0.01</span>))</span>
<span id="cb35-3"><a href="#cb35-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-4"><a href="#cb35-4" aria-hidden="true" tabindex="-1"></a><span class="fu">circos.initialize</span>(<span class="at">factors=</span>chrom_lens<span class="sc">$</span>chr,</span>
<span id="cb35-5"><a href="#cb35-5" aria-hidden="true" tabindex="-1"></a>                  <span class="at">xlim=</span><span class="fu">cbind</span>(chrom_lens<span class="sc">$</span>start, chrom_lens<span class="sc">$</span>end))</span>
<span id="cb35-6"><a href="#cb35-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-7"><a href="#cb35-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Chromosomes at the outside</span></span>
<span id="cb35-8"><a href="#cb35-8" aria-hidden="true" tabindex="-1"></a><span class="fu">circos.trackPlotRegion</span>(</span>
<span id="cb35-9"><a href="#cb35-9" aria-hidden="true" tabindex="-1"></a>  <span class="at">ylim=</span><span class="fu">c</span>(<span class="dv">0</span>,<span class="dv">1</span>),</span>
<span id="cb35-10"><a href="#cb35-10" aria-hidden="true" tabindex="-1"></a>  <span class="at">track.height=</span><span class="fl">0.05</span>,</span>
<span id="cb35-11"><a href="#cb35-11" aria-hidden="true" tabindex="-1"></a>  <span class="at">bg.border=</span><span class="cn">NA</span>,</span>
<span id="cb35-12"><a href="#cb35-12" aria-hidden="true" tabindex="-1"></a>  <span class="at">bg.col =</span> chrom_colors,</span>
<span id="cb35-13"><a href="#cb35-13" aria-hidden="true" tabindex="-1"></a>  <span class="at">panel.fun=</span><span class="cf">function</span>(x,y){</span>
<span id="cb35-14"><a href="#cb35-14" aria-hidden="true" tabindex="-1"></a>    chr <span class="ot">&lt;-</span> CELL_META<span class="sc">$</span>sector.index</span>
<span id="cb35-15"><a href="#cb35-15" aria-hidden="true" tabindex="-1"></a>    <span class="fu">circos.text</span>(</span>
<span id="cb35-16"><a href="#cb35-16" aria-hidden="true" tabindex="-1"></a>      CELL_META<span class="sc">$</span>xcenter,</span>
<span id="cb35-17"><a href="#cb35-17" aria-hidden="true" tabindex="-1"></a>      CELL_META<span class="sc">$</span>ycenter<span class="sc">+</span> <span class="fu">mm_y</span>(<span class="dv">4</span>),</span>
<span id="cb35-18"><a href="#cb35-18" aria-hidden="true" tabindex="-1"></a>      <span class="at">col =</span> <span class="st">"black"</span>,</span>
<span id="cb35-19"><a href="#cb35-19" aria-hidden="true" tabindex="-1"></a>      <span class="at">labels =</span> new_labels[chr],   <span class="co"># Choosing name for each sector</span></span>
<span id="cb35-20"><a href="#cb35-20" aria-hidden="true" tabindex="-1"></a>      <span class="at">facing=</span><span class="st">"inside"</span>,</span>
<span id="cb35-21"><a href="#cb35-21" aria-hidden="true" tabindex="-1"></a>      <span class="at">niceFacing=</span>F,</span>
<span id="cb35-22"><a href="#cb35-22" aria-hidden="true" tabindex="-1"></a>      <span class="at">cex=</span><span class="fl">0.7</span></span>
<span id="cb35-23"><a href="#cb35-23" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb35-24"><a href="#cb35-24" aria-hidden="true" tabindex="-1"></a>    <span class="fu">circos.genomicAxis</span>(<span class="at">h=</span><span class="st">"top"</span>, <span class="at">labels.cex =</span> <span class="fl">0.5</span>)</span>
<span id="cb35-25"><a href="#cb35-25" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb35-26"><a href="#cb35-26" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Note: 1 point is out of plotting region in sector 'A_1', track '1'.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Note: 1 point is out of plotting region in sector 'A_2', track '1'.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Note: 1 point is out of plotting region in sector 'A_3', track '1'.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Note: 1 point is out of plotting region in sector 'A_4', track '1'.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Note: 1 point is out of plotting region in sector 'A_5', track '1'.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Note: 1 point is out of plotting region in sector 'A_6', track '1'.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Note: 1 point is out of plotting region in sector 'B_NZ_CP186561.1',
track '1'.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Note: 1 point is out of plotting region in sector 'B_NZ_CP186562.1',
track '1'.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Note: 1 point is out of plotting region in sector 'B_NZ_CP186563.1',
track '1'.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Note: 1 point is out of plotting region in sector 'B_NZ_CP186564.1',
track '1'.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Note: 1 point is out of plotting region in sector 'B_NZ_CP186565.1',
track '1'.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Note: 1 point is out of plotting region in sector 'B_NZ_CP186566.1',
track '1'.</code></pre>
</div>
<div class="sourceCode cell-code" id="cb48"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb48-1"><a href="#cb48-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Track 1: gene densities</span></span>
<span id="cb48-2"><a href="#cb48-2" aria-hidden="true" tabindex="-1"></a>maxA <span class="ot">&lt;-</span> <span class="fu">max</span>(gendichteA<span class="sc">$</span>value)</span>
<span id="cb48-3"><a href="#cb48-3" aria-hidden="true" tabindex="-1"></a>maxB <span class="ot">&lt;-</span> <span class="fu">max</span>(gendichteB<span class="sc">$</span>value)</span>
<span id="cb48-4"><a href="#cb48-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb48-5"><a href="#cb48-5" aria-hidden="true" tabindex="-1"></a>col_funA <span class="ot">&lt;-</span> <span class="fu">colorRamp2</span>(<span class="fu">c</span>(<span class="dv">0</span>, maxA), <span class="fu">c</span>(<span class="st">"#eeeeee"</span>, <span class="st">"#00a087B2"</span>))  <span class="co"># grey → green</span></span>
<span id="cb48-6"><a href="#cb48-6" aria-hidden="true" tabindex="-1"></a>col_funB <span class="ot">&lt;-</span> <span class="fu">colorRamp2</span>(<span class="fu">c</span>(<span class="dv">0</span>, maxB), <span class="fu">c</span>(<span class="st">"#eeeeee"</span>, <span class="st">"#0c5488b2"</span>))  <span class="co"># grey → blue</span></span>
<span id="cb48-7"><a href="#cb48-7" aria-hidden="true" tabindex="-1"></a>gendichte_df <span class="ot">&lt;-</span> <span class="fu">rbind</span>(gendichteA, gendichteB)</span>
<span id="cb48-8"><a href="#cb48-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb48-9"><a href="#cb48-9" aria-hidden="true" tabindex="-1"></a><span class="fu">circos.track</span>(<span class="at">ylim=</span><span class="fu">c</span>(<span class="dv">0</span>,<span class="dv">1</span>), <span class="at">track.height=</span><span class="fl">0.1</span>, <span class="at">bg.border=</span><span class="cn">NA</span>,</span>
<span id="cb48-10"><a href="#cb48-10" aria-hidden="true" tabindex="-1"></a>             <span class="at">panel.fun=</span><span class="cf">function</span>(region,value,...){</span>
<span id="cb48-11"><a href="#cb48-11" aria-hidden="true" tabindex="-1"></a>               chr <span class="ot">&lt;-</span> CELL_META<span class="sc">$</span>sector.index</span>
<span id="cb48-12"><a href="#cb48-12" aria-hidden="true" tabindex="-1"></a>               idx <span class="ot">&lt;-</span> gendichte_df<span class="sc">$</span>chr <span class="sc">==</span> chr</span>
<span id="cb48-13"><a href="#cb48-13" aria-hidden="true" tabindex="-1"></a>               <span class="cf">for</span>(i <span class="cf">in</span> <span class="fu">which</span>(idx)){</span>
<span id="cb48-14"><a href="#cb48-14" aria-hidden="true" tabindex="-1"></a>                 col_use <span class="ot">&lt;-</span> <span class="fu">ifelse</span>(<span class="fu">grepl</span>(<span class="st">"^A_"</span>, chr), <span class="fu">col_funA</span>(gendichte_df<span class="sc">$</span>value[i]), <span class="fu">col_funB</span>(gendichte_df<span class="sc">$</span>value[i]))</span>
<span id="cb48-15"><a href="#cb48-15" aria-hidden="true" tabindex="-1"></a>                 <span class="fu">circos.rect</span>(gendichte_df<span class="sc">$</span>start[i], <span class="dv">0</span>,</span>
<span id="cb48-16"><a href="#cb48-16" aria-hidden="true" tabindex="-1"></a>                             gendichte_df<span class="sc">$</span>end[i], <span class="dv">1</span>,</span>
<span id="cb48-17"><a href="#cb48-17" aria-hidden="true" tabindex="-1"></a>                             <span class="at">col=</span>col_use, <span class="at">border=</span><span class="cn">NA</span>)</span>
<span id="cb48-18"><a href="#cb48-18" aria-hidden="true" tabindex="-1"></a>               }</span>
<span id="cb48-19"><a href="#cb48-19" aria-hidden="true" tabindex="-1"></a>             })</span>
<span id="cb48-20"><a href="#cb48-20" aria-hidden="true" tabindex="-1"></a><span class="co"># Track 2: Phage sequences</span></span>
<span id="cb48-21"><a href="#cb48-21" aria-hidden="true" tabindex="-1"></a>repeat_types <span class="ot">&lt;-</span> <span class="fu">unique</span>(repeat_df<span class="sc">$</span>type)</span>
<span id="cb48-22"><a href="#cb48-22" aria-hidden="true" tabindex="-1"></a>colors_repeats <span class="ot">&lt;-</span> <span class="fu">setNames</span>(<span class="fu">pal_npg</span>(<span class="st">"nrc"</span>, <span class="at">alpha =</span> <span class="fl">0.7</span>)(<span class="dv">2</span>)[<span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(repeat_types)], repeat_types)</span>
<span id="cb48-23"><a href="#cb48-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb48-24"><a href="#cb48-24" aria-hidden="true" tabindex="-1"></a><span class="fu">circos.track</span>(</span>
<span id="cb48-25"><a href="#cb48-25" aria-hidden="true" tabindex="-1"></a>  <span class="at">ylim =</span> <span class="fu">c</span>(<span class="dv">0</span>,<span class="dv">1</span>), </span>
<span id="cb48-26"><a href="#cb48-26" aria-hidden="true" tabindex="-1"></a>  <span class="at">track.height =</span> <span class="fl">0.1</span>,</span>
<span id="cb48-27"><a href="#cb48-27" aria-hidden="true" tabindex="-1"></a>  <span class="at">bg.border =</span> <span class="st">"grey"</span>, </span>
<span id="cb48-28"><a href="#cb48-28" aria-hidden="true" tabindex="-1"></a>  <span class="at">bg.col =</span> <span class="st">"grey95"</span>,</span>
<span id="cb48-29"><a href="#cb48-29" aria-hidden="true" tabindex="-1"></a>  <span class="at">panel.fun =</span> <span class="cf">function</span>(region, value, ...) {</span>
<span id="cb48-30"><a href="#cb48-30" aria-hidden="true" tabindex="-1"></a>    chr <span class="ot">&lt;-</span> CELL_META<span class="sc">$</span>sector.index</span>
<span id="cb48-31"><a href="#cb48-31" aria-hidden="true" tabindex="-1"></a>    idx <span class="ot">&lt;-</span> repeat_df<span class="sc">$</span>chr <span class="sc">==</span> chr</span>
<span id="cb48-32"><a href="#cb48-32" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb48-33"><a href="#cb48-33" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span>(<span class="fu">any</span>(idx)) {</span>
<span id="cb48-34"><a href="#cb48-34" aria-hidden="true" tabindex="-1"></a>      ybottom <span class="ot">&lt;-</span> CELL_META<span class="sc">$</span>ylim[<span class="dv">1</span>]</span>
<span id="cb48-35"><a href="#cb48-35" aria-hidden="true" tabindex="-1"></a>      ytop    <span class="ot">&lt;-</span> CELL_META<span class="sc">$</span>ylim[<span class="dv">2</span>]</span>
<span id="cb48-36"><a href="#cb48-36" aria-hidden="true" tabindex="-1"></a>      </span>
<span id="cb48-37"><a href="#cb48-37" aria-hidden="true" tabindex="-1"></a>      <span class="fu">circos.rect</span>(</span>
<span id="cb48-38"><a href="#cb48-38" aria-hidden="true" tabindex="-1"></a>        <span class="at">xleft  =</span> repeat_df<span class="sc">$</span>start[idx],</span>
<span id="cb48-39"><a href="#cb48-39" aria-hidden="true" tabindex="-1"></a>        <span class="at">ybottom=</span> ybottom,</span>
<span id="cb48-40"><a href="#cb48-40" aria-hidden="true" tabindex="-1"></a>        <span class="at">xright =</span> repeat_df<span class="sc">$</span>end[idx],</span>
<span id="cb48-41"><a href="#cb48-41" aria-hidden="true" tabindex="-1"></a>        <span class="at">ytop   =</span> ytop,</span>
<span id="cb48-42"><a href="#cb48-42" aria-hidden="true" tabindex="-1"></a>        <span class="at">col    =</span> <span class="fu">adjustcolor</span>(colors_repeats[repeat_df<span class="sc">$</span>type[idx]], <span class="at">alpha.f=</span><span class="fl">0.7</span>),</span>
<span id="cb48-43"><a href="#cb48-43" aria-hidden="true" tabindex="-1"></a>        <span class="at">border =</span> <span class="cn">NA</span></span>
<span id="cb48-44"><a href="#cb48-44" aria-hidden="true" tabindex="-1"></a>      )</span>
<span id="cb48-45"><a href="#cb48-45" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb48-46"><a href="#cb48-46" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb48-47"><a href="#cb48-47" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb48-48"><a href="#cb48-48" aria-hidden="true" tabindex="-1"></a><span class="co"># Track 3: GC content</span></span>
<span id="cb48-49"><a href="#cb48-49" aria-hidden="true" tabindex="-1"></a>gc_df <span class="ot">&lt;-</span> <span class="fu">rbind</span>(gcA,gcB)</span>
<span id="cb48-50"><a href="#cb48-50" aria-hidden="true" tabindex="-1"></a><span class="fu">circos.track</span>(<span class="at">ylim=</span><span class="fu">c</span>(<span class="dv">0</span>,<span class="dv">1</span>), <span class="at">track.height=</span><span class="fl">0.1</span>, <span class="at">bg.border=</span><span class="st">"grey"</span>, <span class="at">bg.col=</span><span class="cn">NA</span>,</span>
<span id="cb48-51"><a href="#cb48-51" aria-hidden="true" tabindex="-1"></a>             <span class="at">panel.fun=</span><span class="cf">function</span>(region,value,...){</span>
<span id="cb48-52"><a href="#cb48-52" aria-hidden="true" tabindex="-1"></a>               chr <span class="ot">&lt;-</span> CELL_META<span class="sc">$</span>sector.index</span>
<span id="cb48-53"><a href="#cb48-53" aria-hidden="true" tabindex="-1"></a>               idx <span class="ot">&lt;-</span> gc_df<span class="sc">$</span>chr <span class="sc">==</span> chr</span>
<span id="cb48-54"><a href="#cb48-54" aria-hidden="true" tabindex="-1"></a>               <span class="fu">circos.lines</span>((gc_df<span class="sc">$</span>start[idx]<span class="sc">+</span>gc_df<span class="sc">$</span>end[idx])<span class="sc">/</span><span class="dv">2</span>, gc_df<span class="sc">$</span>value[idx],</span>
<span id="cb48-55"><a href="#cb48-55" aria-hidden="true" tabindex="-1"></a>                            <span class="at">col=</span><span class="st">"#e64b35b2"</span>, <span class="at">lwd=</span><span class="dv">1</span>)</span>
<span id="cb48-56"><a href="#cb48-56" aria-hidden="true" tabindex="-1"></a>             })</span>
<span id="cb48-57"><a href="#cb48-57" aria-hidden="true" tabindex="-1"></a>first_chr <span class="ot">&lt;-</span> chrom_lens<span class="sc">$</span>chr[<span class="dv">1</span>]</span>
<span id="cb48-58"><a href="#cb48-58" aria-hidden="true" tabindex="-1"></a><span class="fu">circos.yaxis</span>(<span class="at">side=</span><span class="st">"left"</span>, <span class="at">sector.index=</span>first_chr,</span>
<span id="cb48-59"><a href="#cb48-59" aria-hidden="true" tabindex="-1"></a>             <span class="at">labels.cex=</span><span class="fl">0.5</span>, <span class="at">col=</span><span class="st">"darkgrey"</span>, <span class="at">labels.col=</span><span class="st">"black"</span>, <span class="at">tick.length=</span><span class="fu">mm_y</span>(<span class="dv">2</span>))</span>
<span id="cb48-60"><a href="#cb48-60" aria-hidden="true" tabindex="-1"></a><span class="co"># Innermost track: synteny</span></span>
<span id="cb48-61"><a href="#cb48-61" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(synteny)){</span>
<span id="cb48-62"><a href="#cb48-62" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span>(synteny<span class="sc">$</span>chrA[i] <span class="sc">%in%</span> chrom_lens<span class="sc">$</span>chr <span class="sc">&amp;&amp;</span></span>
<span id="cb48-63"><a href="#cb48-63" aria-hidden="true" tabindex="-1"></a>     synteny<span class="sc">$</span>chrB[i] <span class="sc">%in%</span> chrom_lens<span class="sc">$</span>chr){</span>
<span id="cb48-64"><a href="#cb48-64" aria-hidden="true" tabindex="-1"></a>    <span class="fu">circos.link</span>(</span>
<span id="cb48-65"><a href="#cb48-65" aria-hidden="true" tabindex="-1"></a>      <span class="at">sector.index1=</span>synteny<span class="sc">$</span>chrA[i],</span>
<span id="cb48-66"><a href="#cb48-66" aria-hidden="true" tabindex="-1"></a>      <span class="at">point1=</span><span class="fu">c</span>(synteny<span class="sc">$</span>startA[i], synteny<span class="sc">$</span>endA[i]),</span>
<span id="cb48-67"><a href="#cb48-67" aria-hidden="true" tabindex="-1"></a>      <span class="at">sector.index2=</span>synteny<span class="sc">$</span>chrB[i],</span>
<span id="cb48-68"><a href="#cb48-68" aria-hidden="true" tabindex="-1"></a>      <span class="at">point2=</span><span class="fu">c</span>(synteny<span class="sc">$</span>startB[i], synteny<span class="sc">$</span>endB[i]),</span>
<span id="cb48-69"><a href="#cb48-69" aria-hidden="true" tabindex="-1"></a>      <span class="at">col =</span> link_colors[i],</span>
<span id="cb48-70"><a href="#cb48-70" aria-hidden="true" tabindex="-1"></a>      <span class="at">border=</span><span class="cn">NA</span></span>
<span id="cb48-71"><a href="#cb48-71" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb48-72"><a href="#cb48-72" aria-hidden="true" tabindex="-1"></a>  }}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Supplementary_R_plots_files/figure-html/unnamed-chunk-11-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section>
</section>
<section id="synteny-and-phylogenetic-tree-of-18-complete-r.-sphaeroides-genomes" class="level1">
<h1>Synteny and phylogenetic tree of 18 complete <em>R. sphaeroides</em> genomes</h1>
<section id="visualization-of-progressivemauve-output" class="level2">
<h2 class="anchored" data-anchor-id="visualization-of-progressivemauve-output">Visualization of progressiveMauve Output</h2>
<section id="loading-required-packages" class="level4">
<h4 class="anchored" data-anchor-id="loading-required-packages">Loading required packages:</h4>
<div class="cell">
<div class="sourceCode cell-code" id="cb49"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb49-1"><a href="#cb49-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggplot2)  <span class="co"># version 4.0.0</span></span>
<span id="cb49-2"><a href="#cb49-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dplyr)    <span class="co"># version 1.1.4</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>
Attaching package: 'dplyr'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following objects are masked from 'package:data.table':

    between, first, last</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following objects are masked from 'package:GenomicRanges':

    intersect, setdiff, union</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following objects are masked from 'package:Biostrings':

    collapse, intersect, setdiff, setequal, union</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following object is masked from 'package:GenomeInfoDb':

    intersect</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following object is masked from 'package:XVector':

    slice</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following objects are masked from 'package:IRanges':

    collapse, desc, intersect, setdiff, slice, union</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following objects are masked from 'package:S4Vectors':

    first, intersect, rename, setdiff, setequal, union</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following objects are masked from 'package:BiocGenerics':

    combine, intersect, setdiff, union</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following objects are masked from 'package:stats':

    filter, lag</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following objects are masked from 'package:base':

    intersect, setdiff, setequal, union</code></pre>
</div>
<div class="sourceCode cell-code" id="cb61"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb61-1"><a href="#cb61-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(stringr)  <span class="co"># version 1.5.1</span></span>
<span id="cb61-2"><a href="#cb61-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(scales)   <span class="co"># version 1.3.0</span></span>
<span id="cb61-3"><a href="#cb61-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(readxl)   <span class="co"># version 1.4.3</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="function-for-parsing-the-progressivemauve-output" class="level4">
<h4 class="anchored" data-anchor-id="function-for-parsing-the-progressivemauve-output">Function for parsing the <code>progressiveMauve</code> output</h4>
</section>
<section id="some-important-additional-data" class="level4">
<h4 class="anchored" data-anchor-id="some-important-additional-data">Some important additional data</h4>
<ul>
<li><p>Location of the progressiveMauve output (<code>xmfa_file</code>)</p></li>
<li><p>Custom genome order, here as vector (<code>genome_order</code>)</p></li>
<li><p>An excel file listing all contig boundaries (<code>df</code>)</p></li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb62"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb62-1"><a href="#cb62-1" aria-hidden="true" tabindex="-1"></a>xmfa_file <span class="ot">&lt;-</span> <span class="st">"/home/Drives/HDD03_06T_SDE/anna/SyntenyPlotDSMSubH2/genomes/reoriented/C_sph_genomes.xmfa"</span></span>
<span id="cb62-2"><a href="#cb62-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb62-3"><a href="#cb62-3" aria-hidden="true" tabindex="-1"></a>genome_order <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"GCF_001576595"</span>, <span class="st">"GCF_000012905"</span>, <span class="st">"GCF_001685625"</span>,</span>
<span id="cb62-4"><a href="#cb62-4" aria-hidden="true" tabindex="-1"></a>                  <span class="st">"GCF_012647365"</span>, <span class="st">"GCF_003324715"</span>, <span class="st">"GCF_002706325"</span>,</span>
<span id="cb62-5"><a href="#cb62-5" aria-hidden="true" tabindex="-1"></a>                  <span class="st">"GCF_000273405"</span>, <span class="st">"GCF_000269625"</span>, <span class="st">"DSM158"</span>,</span>
<span id="cb62-6"><a href="#cb62-6" aria-hidden="true" tabindex="-1"></a>                  <span class="st">"GCF_049434525"</span>, <span class="st">"GCF_000212605"</span>, <span class="st">'GCF_052246835'</span>,</span>
<span id="cb62-7"><a href="#cb62-7" aria-hidden="true" tabindex="-1"></a>                  <span class="st">"GCF_000021005"</span>, <span class="st">"GCF_003846365"</span>, <span class="st">"GCF_003846425"</span>,</span>
<span id="cb62-8"><a href="#cb62-8" aria-hidden="true" tabindex="-1"></a>                  <span class="st">"GCF_000015985"</span>, <span class="st">"GCF_003846385"</span>, <span class="st">"GCF_003846405"</span>)</span>
<span id="cb62-9"><a href="#cb62-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb62-10"><a href="#cb62-10" aria-hidden="true" tabindex="-1"></a>df <span class="ot">&lt;-</span> <span class="fu">read_excel</span>(<span class="st">"/home/anna/R/Genome_Overview_Rsph.xlsx"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="reading-the-xmfa-file" class="level4">
<h4 class="anchored" data-anchor-id="reading-the-xmfa-file">Reading the <code>xmfa</code> file</h4>
<p>To simplify the visualization only blocks located in two or more assemblies are processed. Therefor we removed all unique blocks (<code>xmfa_data_short</code>).</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb63"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb63-1"><a href="#cb63-1" aria-hidden="true" tabindex="-1"></a>xmfa_data <span class="ot">&lt;-</span> <span class="fu">read_xmfa</span>(xmfa_file)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Number of read lines: 1445608 
Final number of rows in the dataframe: 1998 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb65"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb65-1"><a href="#cb65-1" aria-hidden="true" tabindex="-1"></a>xmfa_data_short <span class="ot">&lt;-</span> xmfa_data[(<span class="dv">1</span><span class="sc">:</span><span class="dv">1683</span>),]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>To plot the different data together, we need to reorder our dataframes (<code>df</code> and <code>xmfa_data_short</code>) after the <code>genome_order</code> vector</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb66"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb66-1"><a href="#cb66-1" aria-hidden="true" tabindex="-1"></a>df<span class="sc">$</span>Genome_Name <span class="ot">&lt;-</span> <span class="fu">factor</span>(</span>
<span id="cb66-2"><a href="#cb66-2" aria-hidden="true" tabindex="-1"></a>  df<span class="sc">$</span>Genome_Name,</span>
<span id="cb66-3"><a href="#cb66-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">levels =</span> genome_order</span>
<span id="cb66-4"><a href="#cb66-4" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb66-5"><a href="#cb66-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb66-6"><a href="#cb66-6" aria-hidden="true" tabindex="-1"></a>xmfa_data_short<span class="sc">$</span>Genome_Name <span class="ot">&lt;-</span> <span class="fu">factor</span>(</span>
<span id="cb66-7"><a href="#cb66-7" aria-hidden="true" tabindex="-1"></a>  xmfa_data_short<span class="sc">$</span>Genome_Name,</span>
<span id="cb66-8"><a href="#cb66-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">levels =</span> genome_order</span>
<span id="cb66-9"><a href="#cb66-9" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="plotting-the-data-with-ggplot2" class="level4">
<h4 class="anchored" data-anchor-id="plotting-the-data-with-ggplot2">Plotting the data with ggplot2</h4>
<div class="cell">
<div class="sourceCode cell-code" id="cb67"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb67-1"><a href="#cb67-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(xmfa_data_short, </span>
<span id="cb67-2"><a href="#cb67-2" aria-hidden="true" tabindex="-1"></a>       <span class="fu">aes</span>(<span class="at">y =</span> Genome_Name)</span>
<span id="cb67-3"><a href="#cb67-3" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">+</span></span>
<span id="cb67-4"><a href="#cb67-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_rect</span>( <span class="co">#LCB from progressiveMauve</span></span>
<span id="cb67-5"><a href="#cb67-5" aria-hidden="true" tabindex="-1"></a>    <span class="fu">aes</span>(<span class="at">xmin =</span> Start, <span class="at">xmax =</span> End,</span>
<span id="cb67-6"><a href="#cb67-6" aria-hidden="true" tabindex="-1"></a>        <span class="at">ymin =</span> <span class="fu">as.numeric</span>(Genome_Name) <span class="sc">-</span> <span class="fl">0.3</span>,</span>
<span id="cb67-7"><a href="#cb67-7" aria-hidden="true" tabindex="-1"></a>        <span class="at">ymax =</span> <span class="fu">as.numeric</span>(Genome_Name) <span class="sc">+</span> <span class="fl">0.3</span>,</span>
<span id="cb67-8"><a href="#cb67-8" aria-hidden="true" tabindex="-1"></a>        <span class="at">fill =</span> <span class="fu">factor</span>(Block))</span>
<span id="cb67-9"><a href="#cb67-9" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">+</span></span>
<span id="cb67-10"><a href="#cb67-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_segment</span>( <span class="co">#Contig boundaries</span></span>
<span id="cb67-11"><a href="#cb67-11" aria-hidden="true" tabindex="-1"></a>    <span class="at">data =</span> df,</span>
<span id="cb67-12"><a href="#cb67-12" aria-hidden="true" tabindex="-1"></a>    <span class="fu">aes</span>(</span>
<span id="cb67-13"><a href="#cb67-13" aria-hidden="true" tabindex="-1"></a>      <span class="at">x =</span> Contig_Boundaries_cum, </span>
<span id="cb67-14"><a href="#cb67-14" aria-hidden="true" tabindex="-1"></a>      <span class="at">xend =</span> Contig_Boundaries_cum,</span>
<span id="cb67-15"><a href="#cb67-15" aria-hidden="true" tabindex="-1"></a>      <span class="at">y =</span> <span class="fu">as.numeric</span>(Genome_Name) <span class="sc">-</span> <span class="fl">0.4</span>,</span>
<span id="cb67-16"><a href="#cb67-16" aria-hidden="true" tabindex="-1"></a>      <span class="at">yend =</span> <span class="fu">as.numeric</span>(Genome_Name) <span class="sc">+</span> <span class="fl">0.4</span></span>
<span id="cb67-17"><a href="#cb67-17" aria-hidden="true" tabindex="-1"></a>    ),</span>
<span id="cb67-18"><a href="#cb67-18" aria-hidden="true" tabindex="-1"></a>    <span class="at">color =</span> <span class="st">"grey20"</span>, </span>
<span id="cb67-19"><a href="#cb67-19" aria-hidden="true" tabindex="-1"></a>    <span class="at">linetype =</span> <span class="st">"dashed"</span>, </span>
<span id="cb67-20"><a href="#cb67-20" aria-hidden="true" tabindex="-1"></a>    <span class="at">linewidth =</span> <span class="fl">0.3</span></span>
<span id="cb67-21"><a href="#cb67-21" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">+</span></span>
<span id="cb67-22"><a href="#cb67-22" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_y_discrete</span>(<span class="at">limits =</span> genome_order) <span class="sc">+</span> <span class="co"># custom genome order</span></span>
<span id="cb67-23"><a href="#cb67-23" aria-hidden="true" tabindex="-1"></a>  <span class="fu">guides</span>(<span class="at">fill =</span> <span class="st">"none"</span>)<span class="sc">+</span> <span class="co">#remove legend</span></span>
<span id="cb67-24"><a href="#cb67-24" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_minimal</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Supplementary_R_plots_files/figure-html/unnamed-chunk-17-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb68"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb68-1"><a href="#cb68-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Finetuning of the figure is made with inkscape</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="generating-a-phylogeny-tree-using-a-distance-matrix-from-mash" class="level4">
<h4 class="anchored" data-anchor-id="generating-a-phylogeny-tree-using-a-distance-matrix-from-mash">Generating a phylogeny tree using a distance matrix from mash</h4>
</section>
<section id="loading-the-required-packages-1" class="level4">
<h4 class="anchored" data-anchor-id="loading-the-required-packages-1">Loading the required packages</h4>
<div class="cell">
<div class="sourceCode cell-code" id="cb69"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb69-1"><a href="#cb69-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ape)        <span class="co"># version 5.7-1</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>
Attaching package: 'ape'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following object is masked from 'package:dplyr':

    where</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following object is masked from 'package:circlize':

    degree</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following object is masked from 'package:Biostrings':

    complement</code></pre>
</div>
<div class="sourceCode cell-code" id="cb74"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb74-1"><a href="#cb74-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)  <span class="co"># version 2.0.0</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>── Attaching core tidyverse packages ──────────────────────── tidyverse 2.0.0 ──
✔ forcats   1.0.0     ✔ readr     2.1.5
✔ lubridate 1.9.3     ✔ tibble    3.2.1
✔ purrr     1.0.2     ✔ tidyr     1.3.1</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>── Conflicts ────────────────────────────────────────── tidyverse_conflicts() ──
✖ lubridate::%within%() masks IRanges::%within%()
✖ dplyr::between()      masks data.table::between()
✖ readr::col_factor()   masks scales::col_factor()
✖ dplyr::collapse()     masks Biostrings::collapse(), IRanges::collapse()
✖ dplyr::combine()      masks BiocGenerics::combine()
✖ purrr::compact()      masks XVector::compact()
✖ dplyr::desc()         masks IRanges::desc()
✖ purrr::discard()      masks scales::discard()
✖ tidyr::expand()       masks S4Vectors::expand()
✖ dplyr::filter()       masks stats::filter()
✖ dplyr::first()        masks data.table::first(), S4Vectors::first()
✖ lubridate::hour()     masks data.table::hour()
✖ lubridate::isoweek()  masks data.table::isoweek()
✖ dplyr::lag()          masks stats::lag()
✖ dplyr::last()         masks data.table::last()
✖ lubridate::mday()     masks data.table::mday()
✖ lubridate::minute()   masks data.table::minute()
✖ lubridate::month()    masks data.table::month()
✖ ggplot2::Position()   masks BiocGenerics::Position(), base::Position()
✖ lubridate::quarter()  masks data.table::quarter()
✖ purrr::reduce()       masks GenomicRanges::reduce(), IRanges::reduce()
✖ dplyr::rename()       masks S4Vectors::rename()
✖ lubridate::second()   masks data.table::second(), S4Vectors::second()
✖ lubridate::second&lt;-() masks S4Vectors::second&lt;-()
✖ dplyr::slice()        masks XVector::slice(), IRanges::slice()
✖ purrr::transpose()    masks data.table::transpose()
✖ lubridate::wday()     masks data.table::wday()
✖ lubridate::week()     masks data.table::week()
✖ ape::where()          masks dplyr::where()
✖ lubridate::yday()     masks data.table::yday()
✖ lubridate::year()     masks data.table::year()
ℹ Use the conflicted package (&lt;http://conflicted.r-lib.org/&gt;) to force all conflicts to become errors</code></pre>
</div>
<div class="sourceCode cell-code" id="cb77"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb77-1"><a href="#cb77-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(reshape2)   <span class="co"># version 1.4.4</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>
Attaching package: 'reshape2'

The following object is masked from 'package:tidyr':

    smiths

The following objects are masked from 'package:data.table':

    dcast, melt</code></pre>
</div>
<div class="sourceCode cell-code" id="cb79"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb79-1"><a href="#cb79-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggplot2)    <span class="co"># version 4.0.0</span></span>
<span id="cb79-2"><a href="#cb79-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggtree)     <span class="co"># version 3.99.0</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>ggtree v3.99.0 Learn more at https://yulab-smu.top/contribution-tree-data/

Please cite:

Guangchuang Yu, Tommy Tsan-Yuk Lam, Huachen Zhu, Yi Guan. Two methods
for mapping and visualizing associated data on phylogeny using ggtree.
Molecular Biology and Evolution. 2018, 35(12):3041-3043.
doi:10.1093/molbev/msy194

Attaching package: 'ggtree'

The following object is masked from 'package:tidyr':

    expand

The following object is masked from 'package:ape':

    rotate

The following object is masked from 'package:Biostrings':

    collapse

The following object is masked from 'package:IRanges':

    collapse

The following object is masked from 'package:S4Vectors':

    expand</code></pre>
</div>
</div>
</section>
<section id="loading-the-output-from-mash" class="level4">
<h4 class="anchored" data-anchor-id="loading-the-output-from-mash">Loading the output from mash:</h4>
<p>It should be a comma-separated file with the following columns:</p>
<ul>
<li><p>Column 1: Name of the first genome</p></li>
<li><p>Column 2: Name of the second genome</p></li>
<li><p>Column 3: Distance between the genomes</p></li>
</ul>
</section>
<section id="then-we-calculated-the-distance-matrix" class="level4">
<h4 class="anchored" data-anchor-id="then-we-calculated-the-distance-matrix">Then we calculated the distance matrix</h4>
<div class="cell">
<div class="sourceCode cell-code" id="cb81"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb81-1"><a href="#cb81-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Mash output: col1=genome1, col2=genome2, col3=distance</span></span>
<span id="cb81-2"><a href="#cb81-2" aria-hidden="true" tabindex="-1"></a>dist_tab <span class="ot">&lt;-</span> <span class="fu">read.table</span>(<span class="st">"/home/Drives/HDD03_06T_SDE/anna/SyntenyPlotDSMSubH2/mash/dist.tab"</span>, </span>
<span id="cb81-3"><a href="#cb81-3" aria-hidden="true" tabindex="-1"></a>                       <span class="at">header=</span><span class="cn">FALSE</span>, </span>
<span id="cb81-4"><a href="#cb81-4" aria-hidden="true" tabindex="-1"></a>                       <span class="at">stringsAsFactors =</span> <span class="cn">FALSE</span>)</span>
<span id="cb81-5"><a href="#cb81-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-6"><a href="#cb81-6" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate distance matrix</span></span>
<span id="cb81-7"><a href="#cb81-7" aria-hidden="true" tabindex="-1"></a>dist_mat <span class="ot">&lt;-</span> reshape2<span class="sc">::</span><span class="fu">acast</span>(dist_tab, V1<span class="sc">~</span>V2, <span class="at">value.var=</span><span class="st">"V3"</span>, <span class="at">fill=</span><span class="dv">0</span>)</span>
<span id="cb81-8"><a href="#cb81-8" aria-hidden="true" tabindex="-1"></a>dist_mat <span class="ot">&lt;-</span> <span class="fu">as.dist</span>(dist_mat)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We used the distance matrix to calculate the neighbor-joining tree and plotted this tree using <code>ggtree</code></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb82"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb82-1"><a href="#cb82-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Neighbor-Joining tree calculation</span></span>
<span id="cb82-2"><a href="#cb82-2" aria-hidden="true" tabindex="-1"></a>tree <span class="ot">&lt;-</span> <span class="fu">nj</span>(dist_mat)</span>
<span id="cb82-3"><a href="#cb82-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-4"><a href="#cb82-4" aria-hidden="true" tabindex="-1"></a><span class="co">#set global option to ignore negative edges</span></span>
<span id="cb82-5"><a href="#cb82-5" aria-hidden="true" tabindex="-1"></a><span class="fu">options</span>(<span class="at">ignore.negative.edge=</span><span class="cn">TRUE</span>)</span>
<span id="cb82-6"><a href="#cb82-6" aria-hidden="true" tabindex="-1"></a><span class="co">#plotting the tree using ggtree</span></span>
<span id="cb82-7"><a href="#cb82-7" aria-hidden="true" tabindex="-1"></a>p_tree <span class="ot">&lt;-</span> <span class="fu">ggtree</span>(tree, <span class="at">branch.length =</span> <span class="st">"none"</span>) <span class="sc">+</span> </span>
<span id="cb82-8"><a href="#cb82-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_tiplab</span>(<span class="at">align =</span> T, <span class="at">as_ylab =</span> T) <span class="sc">+</span></span>
<span id="cb82-9"><a href="#cb82-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">"Phylogenie (Mash-Distanzen)"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Scale for y is already present.
Adding another scale for y, which will replace the existing scale.</code></pre>
</div>
<div class="sourceCode cell-code" id="cb84"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb84-1"><a href="#cb84-1" aria-hidden="true" tabindex="-1"></a>p_tree</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Supplementary_R_plots_files/figure-html/unnamed-chunk-20-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section>
</section>
</section>

</main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->




</body></html>
